### PR TITLE
add timeout and http store in communication

### PR DESCRIPTION
**Pull request by xjqbest**
* add timeout and http store in communication
* add revert and confirm in fleet
* test=develop

### CODE REVIEW

File: paddle/fluid/framework/fleet/gloo_wrapper.cc [hunk 1]
```diff
@@ -35,49 +37,85 @@ void HdfsStore::set(const std::string& key, const std::vector<char>& data) {
...
     fp.reset();
-    break;
+    if (err_no != 0) {
+      VLOG(0) << "fs_open_write failed, retry times " << i << " err no "
+              << err_no;
+      sleep(wait_sleep_ms_ / 1000);
+      paddle::framework::fs_remove(tmp);
+      if (i == retry_times_) {
+        VLOG(0) << "fs_open_write failed, retry times reaches limit";
+        exit(-1);
```
Comment (gongweibao) [inline paddle/fluid/framework/fleet/gloo_wrapper.cc]: Throw exception  and deal with it outside ? or return error code?

**Review by gongweibao — CHANGES_REQUESTED** [review]
PADDLE_WITH_GLOO
用户不设置就没有效果也不报错是吗？

File: paddle/fluid/framework/fleet/gloo_wrapper.cc [hunk 2]
```diff
@@ -118,50 +167,122 @@ std::string HdfsStore::ObjectPath(const std::string& name) {
...
     }
+    (*keys_check_status)[i] = is_exists;
   }
+  return ret;
 #endif
   return true;
 }
 
+#ifdef PADDLE_WITH_GLOO
+void ParallelConnectContext::connectFullMesh(
```
Comment (gongweibao) [inline paddle/fluid/framework/fleet/gloo_wrapper.cc:199]: 函数命名

**Comment by xjqbest** [comment]
> PADDLE_WITH_GLOO
> 用户不设置就没有效果也不报错是吗？

add some warnings, thanks

File: paddle/fluid/framework/fleet/gloo_wrapper.cc (hunk 1 above)
Comment (xjqbest) [reply paddle/fluid/framework/fleet/gloo_wrapper.cc]: add exception, thanks

File: paddle/fluid/framework/fleet/gloo_wrapper.cc (hunk 2 above)
Comment (xjqbest) [reply paddle/fluid/framework/fleet/gloo_wrapper.cc:199]: override base class fucntion, so the name is like that.

**Review by fuyinno4 — DISMISSED** [review]

File: paddle/fluid/framework/fleet/fleet_wrapper.cc [hunk 3]
```diff
@@ -1166,6 +1166,24 @@ int32_t FleetWrapper::CopyTable(const uint64_t src_table_id,
 #endif
 }
 
+void FleetWrapper::Confirm() {
+#ifdef PADDLE_WITH_PSLIB
+// auto ret = pslib_ptr_->_worker_ptr->confirm();
```
Comment (gongweibao) [inline paddle/fluid/framework/fleet/fleet_wrapper.cc]: Cleanup?

File: paddle/fluid/framework/fleet/gloo_wrapper.cc [hunk 4]
```diff
@@ -35,49 +37,88 @@ void HdfsStore::set(const std::string& key, const std::vector<char>& data) {
...
   paddle::framework::fs_mv(tmp, path);
 #endif
 }
 
+#ifdef PADDLE_WITH_GLOO
+int retry_do_func(std::function<int(void)> func, uint32_t max_try_time,
+                  uint32_t retry_interval_ms) {
+  int ret = -1;
+  for (uint32_t i = 0; i < max_try_time; ++i) {
+    ret = func();
```
Comment (gongweibao) [inline paddle/fluid/framework/fleet/gloo_wrapper.cc]:
```
if(func == 0){
    break;
}
```

File: paddle/fluid/framework/fleet/gloo_wrapper.cc [hunk 5]
```diff
@@ -35,49 +37,88 @@ void HdfsStore::set(const std::string& key, const std::vector<char>& data) {
...
+    if (ret == 0) {
+      break;
+    }
+#ifdef _LINUX
```
Comment (gongweibao) [inline paddle/fluid/framework/fleet/gloo_wrapper.cc:77]: else?

File: paddle/fluid/framework/fleet/gloo_wrapper.cc [hunk 6]
```diff
@@ -35,49 +37,88 @@ void HdfsStore::set(const std::string& key, const std::vector<char>& data) {
...
     fp.reset();
-    break;
+    if (err_no != 0) {
+      VLOG(0) << "fs_open_write failed, retry times " << i << " err no "
+              << err_no;
+      sleep(wait_sleep_ms_ / 1000);
+      paddle::framework::fs_remove(tmp);
+      if (i == retry_times_) {
+        VLOG(0) << "fs_open_write failed, retry times reaches limit";
+        PADDLE_ENFORCE_EQ(0, 1, platform::errors::PreconditionNotMet(
```
Comment (gongweibao) [inline paddle/fluid/framework/fleet/gloo_wrapper.cc]: Throw?

File: paddle/fluid/framework/fleet/gloo_wrapper.cc [hunk 7]
```diff
@@ -92,22 +133,33 @@ void HdfsStore::wait(const std::vector<std::string>& keys,
...
-                                  "HdfsStore::wait, Wait timeout for key(s): " +
-                                  ::gloo::MakeString(keys)));
+      int32_t last_check_rank = -1;
+      for (size_t i = 0; i < check_key_status.size(); ++i) {
+        if (!check_key_status[i]) {
+          last_check_rank = i;
+          break;
+        }
+      }
+      PADDLE_ENFORCE_EQ(0, 1, platform::errors::ExecutionTimeout(
```
Comment (gongweibao) [inline paddle/fluid/framework/fleet/gloo_wrapper.cc]: Throw?

File: python/paddle/fluid/tests/unittests/test_fleet_rolemaker_4.py [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+                def __init__(self):
+                    """Init."""
+                    self.delete_kv_lock = threading.Lock()
+                    self.delete_kv = {}
+                    self.kv_lock = threading.Lock()
+                    self.kv = {}
+        except:
+            print("warning: no KVHTTPServer, skip test_pslib_4")
+            return
+
```
Comment (gongweibao) [inline python/paddle/fluid/tests/unittests/test_fleet_rolemaker_4.py:155]: 为何间隔这么多？

**Review by gongweibao — CHANGES_REQUESTED** [review]

File: paddle/fluid/framework/fleet/gloo_wrapper.cc [hunk 9]
```diff
@@ -118,50 +170,122 @@ std::string HdfsStore::ObjectPath(const std::string& name) {
   return path_ + "/" + EncodeName(name);
 }
 
-bool HdfsStore::Check(const std::vector<std::string>& keys) {
+bool HdfsStore::Check(const std::vector<std::string>& keys,
+                      std::vector<bool>* keys_check_status) {
 #ifdef PADDLE_WITH_GLOO
```
Comment (gongweibao) [inline paddle/fluid/framework/fleet/gloo_wrapper.cc:173]: else?

**Review by gongweibao — CHANGES_REQUESTED** [review]

File: python/paddle/fluid/tests/unittests/test_fleet_rolemaker_4.py (hunk 8 above)
Comment (xjqbest) [reply python/paddle/fluid/tests/unittests/test_fleet_rolemaker_4.py:155]: 代码风格ci导致的，如果不加过不了ci～

File: paddle/fluid/framework/fleet/gloo_wrapper.cc (hunk 7 above)
Comment (xjqbest) [reply paddle/fluid/framework/fleet/gloo_wrapper.cc]: fixed, thx

File: paddle/fluid/framework/fleet/gloo_wrapper.cc (hunk 6 above)
Comment (xjqbest) [reply paddle/fluid/framework/fleet/gloo_wrapper.cc]: fixed, thx

File: paddle/fluid/framework/fleet/gloo_wrapper.cc (hunk 5 above)
Comment (xjqbest) [reply paddle/fluid/framework/fleet/gloo_wrapper.cc:77]: fixed, thx

File: paddle/fluid/framework/fleet/gloo_wrapper.cc (hunk 4 above)
Comment (xjqbest) [reply paddle/fluid/framework/fleet/gloo_wrapper.cc]: fixed, thx

File: paddle/fluid/framework/fleet/fleet_wrapper.cc (hunk 3 above)
Comment (xjqbest) [reply paddle/fluid/framework/fleet/fleet_wrapper.cc]: fixed, thx

File: paddle/fluid/framework/fleet/gloo_wrapper.cc (hunk 9 above)
Comment (xjqbest) [reply paddle/fluid/framework/fleet/gloo_wrapper.cc:173]: fixed, thx

File: paddle/fluid/framework/fleet/gloo_wrapper.cc [hunk 10]
```diff
@@ -35,49 +37,86 @@ void HdfsStore::set(const std::string& key, const std::vector<char>& data) {
...
     }
     fp.reset();
-    break;
+    if (err_no != 0) {
+      VLOG(0) << "fs_open_write failed, retry times " << i << " err no "
+              << err_no;
+      sleep(wait_sleep_ms_ / 1000);
+      paddle::framework::fs_remove(tmp);
+      if (i == retry_times_) {
+        VLOG(0) << "fs_open_write failed, retry times reaches limit";
```
Comment (gongweibao) [inline paddle/fluid/framework/fleet/gloo_wrapper.cc:56]: log error

File: paddle/fluid/framework/fleet/gloo_wrapper.cc (hunk 5 above)
Comment (gongweibao) [reply paddle/fluid/framework/fleet/gloo_wrapper.cc:77]: ?

File: paddle/fluid/framework/fleet/gloo_wrapper.cc [hunk 11]
```diff
@@ -118,50 +168,124 @@ std::string HdfsStore::ObjectPath(const std::string& name) {
...
     VLOG(3) << "HdfsStore::Check " << is_exists << " path " << path;
     if (!is_exists) {
-      return false;
+      ret = false;
     }
+    (*keys_check_status)[i] = is_exists;
   }
+  return ret;
+#else
+  VLOG(0) << "HdfsStore::Check does nothing when no gloo";
```
Comment (gongweibao) [inline paddle/fluid/framework/fleet/gloo_wrapper.cc:193]: throw exception.

File: paddle/fluid/framework/fleet/gloo_wrapper.cc [hunk 12]
```diff
@@ -118,50 +168,124 @@ std::string HdfsStore::ObjectPath(const std::string& name) {
...
+          http_ip_, http_port_, prefix_ + "_" + http_scope_, rank_);
+      http_store->SetTimeoutSeconds(init_timeout_.count());
+      context->connectFullMesh(*http_store, dev);
+      http_store->Finalize();
+      break;
+    }
+    default:
+      LOG(ERROR) << "unknown store type " << store_type_;
+      exit(-1);
+  }
```
Comment (gongweibao) [inline paddle/fluid/framework/fleet/gloo_wrapper.cc:288]: else throw exception.

File: paddle/fluid/framework/fleet/gloo_wrapper.cc [hunk 13]
```diff
@@ -118,50 +168,124 @@ std::string HdfsStore::ObjectPath(const std::string& name) {
...
+    case GlooStoreType::HTTP: {
+      http_store = std::make_shared<gloo::rendezvous::HTTPStore>(
+          http_ip_, http_port_, prefix_ + "_" + http_scope_, rank_);
+      http_store->SetTimeoutSeconds(init_timeout_.count());
+      context->connectFullMesh(*http_store, dev);
+      http_store->Finalize();
+      break;
+    }
+    default:
+      LOG(ERROR) << "unknown store type " << store_type_;
```
Comment (gongweibao) [inline paddle/fluid/framework/fleet/gloo_wrapper.cc:286]: throw exception!

File: paddle/fluid/framework/fleet/gloo_wrapper.h [hunk 14]
```diff
@@ -147,6 +195,8 @@ class GlooWrapper {
     opts.setInput(&input, 1);
     opts.setOutput(ret.data(), size_);
     gloo::allgather(opts);
+#else
+    LOG(WARNING) << "AllGather does nothing when WITH_GLOO=OFF";
```
Comment (gongweibao) [inline paddle/fluid/framework/fleet/gloo_wrapper.h:199]: throw exception.

File: paddle/fluid/string/string_helper.cc [hunk 15]
```diff
@@ -61,6 +61,19 @@ std::string trim_spaces(const std::string& str) {
   return std::string(p, len);
 }
 
+std::string erase_spaces(const std::string& str) {
```
Comment (gongweibao) [inline paddle/fluid/string/string_helper.cc:64]: https://stackoverflow.com/questions/83439/remove-spaces-from-stdstring-in-c

**Review by gongweibao — CHANGES_REQUESTED** [review]

**Review by gongweibao — CHANGES_REQUESTED** [review]
调用功能需要给用户明确的反应！

File: paddle/fluid/string/string_helper.cc (hunk 15 above)
Comment (xjqbest) [reply paddle/fluid/string/string_helper.cc:64]: will fix in next pr

File: paddle/fluid/framework/fleet/gloo_wrapper.h (hunk 14 above)
Comment (xjqbest) [reply paddle/fluid/framework/fleet/gloo_wrapper.h:199]: will fix in next pr

File: paddle/fluid/framework/fleet/gloo_wrapper.cc (hunk 13 above)
Comment (xjqbest) [reply paddle/fluid/framework/fleet/gloo_wrapper.cc:286]: will fix in next pr

File: paddle/fluid/framework/fleet/gloo_wrapper.cc (hunk 12 above)
Comment (xjqbest) [reply paddle/fluid/framework/fleet/gloo_wrapper.cc:288]: will fix in next pr

File: paddle/fluid/framework/fleet/gloo_wrapper.cc (hunk 11 above)
Comment (xjqbest) [reply paddle/fluid/framework/fleet/gloo_wrapper.cc:193]: will fix in next pr

File: paddle/fluid/framework/fleet/gloo_wrapper.cc (hunk 10 above)
Comment (xjqbest) [reply paddle/fluid/framework/fleet/gloo_wrapper.cc:56]: will fix in next pr

**Comment by xjqbest** [comment]
> 调用功能需要给用户明确的反应！

will fix in next pr

**Review by gongweibao — APPROVED** [review]
LGTM
还有很多问题下个PR请修复！

**Comment by xjqbest** [comment]
> LGTM
> 还有很多问题下个PR请修复！

收到

**Pull request merged**